### PR TITLE
refactor: Add AccessList result type to eth_createAccessList

### LINKED ISSUE #9746: Add AccessList result type to eth_createAccessList
### Describe the feature

eth_createAccesslist returns a special result type

https://github.com/ethereum/go-ethereum/blob/380688c636a654becc8f114438c2a5d93d2db032/internal/ethapi/api.go#L1475-L1482

where the error is optional and is the evm error

https://github.com/ethereum/go-ethereum/blob/380688c636a654becc8f114438c2a5d93d2db032/internal/ethapi/api.go#L1491-L1498

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by c0np4nn4**
close #9746.

@mattsse I made changes based on the conversation we had on the issue #9746 .
I wonder is this the right way to handle the issue. Thank you.

### CODE REVIEW

File: crates/primitives/src/transaction/access_list.rs [hunk 1]
```diff
@@ -3,6 +3,19 @@
...
+
+/// `AccessListWithGasUsedAndError` for handling error from `eth_createAccessList`
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AccessListWithGasUsedAndError {
+    /// List with accounts accessed during transaction.
+    pub access_list: AccessList,
+    /// Estimated gas used with access list.
+    pub gas_used: U256,
+    /// Optional error message if the transaction failed.
+    pub error: Option<String>,
```
Comment (mattsse) [inline crates/primitives/src/transaction/access_list.rs:17]: this should be skipped if none

 #[serde(default, skip_serializing_if = "Option::is_none")]

File: crates/rpc/rpc-eth-api/src/helpers/call.rs [hunk 2]
```diff
@@ -239,18 +239,19 @@ pub trait EthCall: Call + LoadPendingBlock {
...
             }
-            ExecutionResult::Success { .. } => Ok(()),
-        }?;
+        };
+
+        let error = match result.result {
+            ExecutionResult::Halt { reason, .. } => Some(format!("{:?}", reason)),
+            ExecutionResult::Revert { output, .. } => Some(format!("{:?}", output)),
+            ExecutionResult::Success { .. } => None,
+        };
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/call.rs:254]: we can keep the code on main and use RevertError::new(output) for the revert case

File: crates/primitives/src/transaction/access_list.rs [hunk 3]
```diff
@@ -3,6 +3,19 @@
 /// Re-export from `alloy_eips`.
 #[doc(inline)]
 pub use alloy_eips::eip2930::{AccessList, AccessListItem};
+use revm_primitives::U256;
+use serde::{Deserialize, Serialize};
+
+/// `AccessListWithGasUsedAndError` for handling error from `eth_createAccessList`
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AccessListWithGasUsedAndError {
```
Comment (mattsse) [inline crates/primitives/src/transaction/access_list.rs:11]: let's also name this AccessListResult

Comment (mattsse) [inline crates/primitives/src/transaction/access_list.rs:11]: this needs camelCase rename

**Review by mattsse — CHANGES_REQUESTED** [review]
ty, some suggestions

File: crates/rpc/rpc-eth-api/src/helpers/call.rs (hunk 2 above)
Comment (c0np4nn4) [reply crates/rpc/rpc-eth-api/src/helpers/call.rs]: I will use the code regarding `(result, env)` from the main branch as is. 
Also, I made some changes to the Revert part and left a commit for it.

**Review by onbjerg — APPROVED** [review]
otherwise lgtm

File: crates/rpc/rpc-eth-api/src/helpers/call.rs [hunk 4]
```diff
@@ -240,16 +240,13 @@ pub trait EthCall: Call + LoadPendingBlock {
...
-        let (result, env) = match self.inspect(&mut db, env, &mut inspector) {
-            Ok((res, env)) => (res, env),
-            Err(e) => {
-                return Err(e.into());
-            }
-        };
+        let (result, env) = self.inspect(&mut db, env, &mut inspector)?;
 
         let error = match result.result {
             ExecutionResult::Halt { reason, .. } => Some(format!("{:?}", reason)),
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/call.rs:246]: this should use + `to_string`

https://github.com/paradigmxyz/reth/blob/ef84899410b1327843721a00d8bf6cf6155cbcb7/crates/rpc/rpc-eth-types/src/error.rs#L410-L410

File: crates/rpc/rpc-eth-api/src/helpers/call.rs [hunk 5]
```diff
@@ -240,16 +240,13 @@ pub trait EthCall: Call + LoadPendingBlock {
...
-            ExecutionResult::Revert { output, .. } => Some(format!("{:?}", output)),
+            ExecutionResult::Revert { output, .. } => {
+                Some(format!("{:?}", RevertError::new(output)))
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/call.rs:248]:
```suggestion
                Some(RevertError::new(output).to_string())
```

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits

File: crates/rpc/rpc-eth-api/src/helpers/call.rs (hunk 4 above)
Comment (c0np4nn4) [reply crates/rpc/rpc-eth-api/src/helpers/call.rs]: @mattsse 

https://github.com/paradigmxyz/reth/blob/d7112927cf617220414f290f5ec233660f431a89/crates/rpc/rpc-eth-api/src/helpers/call.rs#L245-L249

Comment (mattsse) [reply crates/rpc/rpc-eth-api/src/helpers/call.rs]: we still need to use 

https://github.com/paradigmxyz/reth/blob/ef84899410b1327843721a00d8bf6cf6155cbcb7/crates/rpc/rpc-eth-types/src/error.rs#L410-L410

to create the appropriate error message

Comment (c0np4nn4) [reply crates/rpc/rpc-eth-api/src/helpers/call.rs]: Sorry to bother you. I fix it.

https://github.com/paradigmxyz/reth/blob/49ef67974604e93395c4be0ae1fbf627466a6f4f/crates/rpc/rpc-eth-api/src/helpers/call.rs#L245-L251

**Review by mattsse — APPROVED** [review]
ty, I changed it slightly to exit right away if the tx failed with the correct gas used and error message

**Pull request merged**